### PR TITLE
fix: Stack overflow on multi maps (and other non-trivial generic clas…

### PR DESCRIPTION
**Pull request by xRodney**
## Description
…ses), schema for multimaps is wrong

Logic in sundrio that is responsible for replacing generic type arguments on methods and properties with their concrete instantiations from subclasses contained a bug, in which it looped infinitely if the same type argument name was used at multiple classes in the hierarchy. A common occurence were multimaps.

A very similar code was also present in crd-generator, also suffering from the same bug.

Sundrio has been updated to account for this situation, and updated utilities from sundrio have been introduced to crd-generator, which replace the previous implementation of the same algorithm.

A test has been added to validate that multimaps (like `class MultiMap<K,V> implements Map<K,List<V>>`) are generated correctly.

Fixes fabric8io/kubernetes-client#4357
Fixes fabric8io/kubernetes-client#4487
Supersedes closes #4504 

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by xRodney** [comment]
It seems that sundrio 0.93.1 is not yet fully released. I will make it Ready for review once it is.

**Comment by xRodney** [comment]
Sundrio has been released (thanks @iocanel), this is now ready for review.
(That one failing E2E test is a false alarm, in my opinion)

**Review by andreaTP — COMMENTED** [review]
Overall this looks good, thanks a lot @xRodney !

I added one question and if you can find more tests to cover possible edge cases it would be nice (e.g. different number of generic type parameters etc. etc.).

File: crd-generator/api/src/test/java/io/fabric8/crd/example/map/ContainingMapsSpec.java [hunk 1]
```diff
@@ -32,4 +33,8 @@ public Map<String, Map<String, List<Boolean>>> getTest2() {
     return test2;
   }
 
+  private MyMultiMap<String, String> test3;
+
+  class MyMultiMap<K, V> extends HashMap<K, List<V>> {
```
[the remark this answers is not in the document]
Comment (xRodney) [reply crd-generator/api/src/test/java/io/fabric8/crd/example/map/ContainingMapsSpec.java:38]: [The new implementation](https://github.com/sundrio/sundrio/blob/main/model/utils/src/main/java/io/sundr/model/functions/TypeCast.java) investigates the whole inheritance tree to see what type arguments are used on `Map`. 

In your example, the parameters on any reference are ignored and the actual type used is always `Map<Integer, Boolean>`, so that's what is rendered. 
(Except ` WARN Property 'test3' with 'io.fabric8.crd.example.map.ContainingMapsSpec.MyMultiMap<java.lang.String,java.lang.String>' key type is mapped to 'string' because of CRD schemas limitations` :D)

```

  private MyMultiMap<String, String> test3;

  class MyMultiMap<K, V> extends HashMap<Integer, Boolean> {
  }
```

becomes:
```
test3:
  additionalProperties:
    type: boolean
  type: object
```

I will add test covering this and other cases as suggested.

**Comment by xRodney** [comment]
I have added more tests to `crd-generator-api`. I also copied most of the cases to `crd-generator-test`, because I was worried about possible differences between Sundrio Reflection and APT contexts. But all seems to be good.

Btw, there is already quite an extensive test suite in sundrio [here](https://github.com/sundrio/sundrio/blob/main/model/utils/src/test/java/io/sundr/model/functions/TypeCastTest.java) and [here](https://github.com/sundrio/sundrio/blob/main/model/utils/src/test/java/io/sundr/model/utils/CollectionsTest.java)

File: crd-generator/test/src/test/java/io/fabric8/crd/generator/maps/ContainingMapsCRDTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import io.fabric8.kubernetes.api.model.apiextensions.v1.JSONSchemaProps;
+import io.fabric8.kubernetes.client.utils.Serialization;
+import org.junit.jupiter.api.Test;
+
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+class ContainingMapsCRDTest {
```
Comment (manusa) [inline crd-generator/test/src/test/java/io/fabric8/crd/generator/maps/ContainingMapsCRDTest.java:29]: Most of these tests could be refactored to use single assertions with AssertJ, but it's not big deal

Comment (manusa) [reply crd-generator/test/src/test/java/io/fabric8/crd/generator/maps/ContainingMapsCRDTest.java:29]: OK, this is probably a port of `CRDGeneratorTest`. That one needs some refactoring too :sweat_smile:

**Review by manusa — APPROVED** [review]
LGTM (codewise)

**Review by andreaTP — APPROVED** [review]
Lgtm! Thanks a lot for this fix @xRodney ! Much appreciated!
